components/brave_shields/browser/brave_shields_web_contents_observer.cc

```diff
@@ -250,4 +261,58 @@ void BraveShieldsWebContentsObserver::RegisterProfilePrefs(
   registry->RegisterUint64Pref(kFingerprintingBlocked, 0);
 }
 
+void BraveShieldsWebContentsObserver::ReadyToCommitNavigation(
+    content::NavigationHandle* navigation_handle) {
+  auto frame_tree_node_id = navigation_handle->GetFrameTreeNodeId();
+  auto *frame_tree_node = content::FrameTreeNode::GloballyFindByID(
+      frame_tree_node_id);
+  auto* navigation_entry =
+    frame_tree_node->navigator()->GetController()->GetPendingEntry();
+
+  if (!navigation_entry) {
+    navigation_entry =
+      frame_tree_node->navigator()->GetController()->GetLastCommittedEntry();
+  }
+  GURL target_origin = navigation_handle->GetURL().GetOrigin();
+  if (!navigation_entry) {
+    return;
+  }
+  Profile* profile =
+      Profile::FromBrowserContext(web_contents()->GetBrowserContext());
+  GURL tab_origin(navigation_entry->GetURL().GetOrigin());
+
+  std::unique_ptr<base::Value> referrer_value =
```

Comment:
seems like we have some duplicated logic with `ApplyPotentialReferrerBlock` that could get out-of-sync. Maybe we should just return the updated (possibly unchanged) referrer so we can use it in both places?

Reply:
This was refactored, but comment is old.